### DISCUSSION TITLE
Global preinstall/postinstall hooks

### DISCUSSION THREAD

**Discussion by knu** [jdx/mise#3844] [Q&A]
It seems the current experimental hooks feature is limited to per-project configuration, but there’s also a need for global preinstall/postinstall hooks.  For instance, when installing a new Rust version, I’d like to automatically place a default Cargo configuration file (config.toml) in its .cargo/ directory to make sure additional compiler flags are used.  When installing a new version of Go or Ruby as an upgrade to the current system-wide default version, I’d like to install some must-have utilities using `go install`/`gem install`.  When installing a new version of PostgreSQL, you could dump the databases of the current version in preinstall and then restore them in postinstall.

These hooks would streamline routine setup tasks and ensure consistency across all versions of tools used in day-to-day development.

**Comment by skatkov** [comment]
It would also be great to pre-install some gems for very ruby managed by mise. example would be `ruby-lsp` gem.

usually it's not expected to add such gems into project's Gemfile.

**Reply by stormwatch** [reply]
Doesn't mise honor the packages listed in `$HOME/.default-gems` like asdf?

**Comment by jdx** [comment]
> It seems the current experimental hooks feature is limited to per-project configuration, but there’s also a need for global preinstall/postinstall hooks.

is this true? it should work in global configs I think

**Reply by stormwatch** [reply]
I believe it does.

**Comment by r1m** [comment]
I would like to set up a postinstall hook to add a custom Cacert into every newly installed JDK but it seems that the preinstall hooks are not run if defined globally in  ~/.config/mise/config.toml

**Reply by stormwatch** [reply]
Is it a post or pre install hook? The [[Tool Options]](https://mise.jdx.dev/dev-tools/#tool-options) documentation, only talks about `postinstall` and doesn't mention `preinstall`.

I confirm that `postinstall` hooks are run for global tools and that a `preinstall` field in a configuration like:
```toml
java = { version = "latest", postinstall = "java-post-install", preinstall = "java-pre-install" }
``` 
is ignored

**Reply by joehorsnell** [reply]
> I would like to set up a postinstall hook to add a custom Cacert into every newly installed JDK

We have exactly this use case too.

The current [Tool Options](https://mise.jdx.dev/dev-tools/#tool-options) config (example copied below) requires a version to be specified when defining a `postinstall` hook.

```
[tools]
node = { version = '20', postinstall = 'corepack enable' }
```

It would be really handy if you could omit the version, so that could vary and be specified per-project, then be combined/merged with the global config that defines the `postinstall` hook.

Alternatively, if not omitted (for the version), then maybe allow a wildcard or pattern to match version numbers?

**Reply by joehorsnell** [reply]
Linking to #4562 as the global hooks was one of the use cases I envisaged for sharing Mise config across repos.
